services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,580 @@
+    <- do
+      len <- let l = length _meRecipients in choose (l `div` 2, l)
+      vectorOf len genProtoAddress
+
+  let addrs :: [(Recipient, Address "no-keys")]
+      addrs = mconcat $ zipWith go _meRecipients protoaddrs
+        where go rcp@(Recipient uid _ cids) adr = (\cid -> (rcp, adr uid cid)) <$> cids
+
+      _meNativeAddress :: Map UserId (Map ClientId (Address "no-keys"))
+      _meNativeAddress = foldl' go mempty (snd <$> addrs)
+        where
+          go :: Map UserId (Map ClientId (Address "no-keys"))
+             -> Address "no-keys"
+             -> Map UserId (Map ClientId (Address "no-keys"))
+          go m addr = Map.alter (go' addr) (addr ^. addrUser) m
+
+          go' :: Address "no-keys"
+              -> Maybe (Map ClientId (Address "no-keys"))
+              -> Maybe (Map ClientId (Address "no-keys"))
+          go' addr = Just . maybe newEntries (newEntries <>)
+            where newEntries = Map.fromList [(addr ^. addrClient, addr)]
+
+  _meWSReachable <- genPredicate . mconcat $ recipientToIds <$> _meRecipients
+  _meNativeReachable <- genPredicate (snd <$> addrs)
+
+  let env = MockEnv {..}
+  validateMockEnv env & either error (const $ pure env)
+
+shrinkMockEnv :: MockEnv -> [MockEnv]
+shrinkMockEnv env = do
+  rcps :: [Recipient] <- shrinkList (const []) (env ^. meRecipients)
+
+  let env' :: MockEnv
+      env' = env
+        & meRecipients .~ rcps
+        & meNativeAddress %~ naddrs
+        & meWSReachable %~ wsrchbl
+        & meNativeReachable %~ ntrchbl
+
+      clients :: Set (UserId, ClientId)
+      clients = Set.fromList . mconcat $ (\(Recipient uid _ cids) -> (uid,) <$> cids) <$> rcps
+
+      naddrs  = Map.filterWithKey (\uid _ -> uid `Set.member` (fst `Set.map` clients))
+              . fmap (Map.filterWithKey (\cid _ -> cid `Set.member` (snd `Set.map` clients)))
+      wsrchbl = Set.filter (`Set.member` clients)
+      ntrchbl = Set.filter (`Set.member` flatten (env' ^. meNativeAddress))
+        where flatten = Set.fromList . mconcat . Map.elems . fmap Map.elems
+
+  [env' | not $ null rcps]
+
+validateMockEnv :: MockEnv -> Either String ()
+validateMockEnv env = do
+  forM_ (Map.toList $ env ^. meNativeAddress) $ \(uid, el) -> do
+    forM_ (Map.toList el) $ \(cid, adr) -> do
+      unless (uid == adr ^. addrUser && cid == adr ^. addrClient) $ do
+        throwError (show (uid, cid, adr))
+
+recipientToIds :: Recipient -> [(UserId, ClientId)]
+recipientToIds (Recipient uid _ cids) = (uid,) <$> cids
+
+genPredicate :: forall a. (Eq a, Ord a) => [a] -> Gen (Set a)
+genPredicate xs = Set.fromList <$> do
+  bools :: [Bool] <- vectorOf (length xs) arbitrary
+  let reachables :: [a] = mconcat $ zipWith (\x yes -> [ x | yes ]) xs bools
+  pure reachables
+
+genRecipient :: HasCallStack => Gen Recipient
+genRecipient = do
+  uid  <- genId
+  cids <- nub <$> listOf1 genClientId
+  pure $ Recipient uid RouteAny cids
+
+genId :: Gen (Id a)
+genId = do
+  gen <- mkStdGen <$> resize 100 arbitrary
```

Comment:
Not sure about the semantics of quickcheck's `resize`, but I think it's less directly related to the 'Gen' value range.  Come to think of it, I think `100` is the default and this is bogus.  Would it be a problem if there were duplicates here?  I'm not sure, but I actually think that would be a good thing for the tests to run into.

Reply:
For integrals, `resize 100 arbitrary` will return numbers in `[-100;100]`.

If we want tests to run into duplicates, it would be better to build that explicitly by doing something like

```haskell
frequency
  [ (1, pure <some ID>)
  , (9, <generate completely random ID>)
  ]
```

Or just use `choose (0, 100)` or something, this one has more immediately understandable semantics (I think).